src/listener.js

```diff
@@ -1,183 +1,234 @@
+        log.error('inbound connection failed to upgrade', err)
+        return maConn.close()
+      }
 
-    conn = await prom(cb => muxed.once('stream', s => cb(null, s)))
-    conn = await this.client.switch.negotiateProtocol(conn, '/p2p/stardust/0.1.0')
-    const rpc = LP(conn)
+      log('inbound connection %s upgraded', maConn.remoteAddr)
+      this.handler && this.handler(conn)
+      this.emit('connection', conn)
+    })
 
-    log('performing challenge')
+    this._onDiscovery()
+  }
 
-    const random = crypto.randomBytes(128)
-    rpc.writeProto(JoinInit, { random128: random, peerID: this.client.id.toJSON() })
+  /**
+   * Read discovery messages
+   * @private
+   */
+  async _onDiscovery () {
+    if (!this.isConnected) {
+      return
+    }
 
-    log('sent rand')
+    const baseAddr = this.address.decapsulate('p2p-stardust')
+    let message
 
-    const { error, saltEncrypted } = await rpc.readProto(JoinChallenge)
-    if (error) { translateAndThrow(error) }
-    const saltSecret = await prom(cb => this.client.id.privKey.decrypt(saltEncrypted, cb))
+    try {
+      message = await this.wrappedStream.readPB(Discovery)
 
-    const solution = sha5(random, saltSecret)
-    rpc.writeProto(JoinChallengeSolution, { solution })
+      // Proof still connected
+      this.wrappedStream.writePB({}, DiscoveryAck)
+    } catch (err) {
+      log('failed to read discovery: %s', err.stack)
+      log('assume disconnected!')
 
-    const { error: error2 } = await rpc.readProto(JoinVerify)
-    if (error2) { translateAndThrow(error2) }
+      this.isConnected = false
+      this.serverConnection = null
+      this.wrappedStream = null
 
-    log('connected')
+      log('reconnecting')
 
-    this.connected = true
-    this.muxed = muxed
-    this.rpc = rpc
-    this.client.connections[this.aid] = this
+      try {
+        await this.listen(this.address)
+        log('reconnected!')
+      } catch (e) {
+        log('reconnect failed: %s', e.stack)
+      }
+
+      return
+    }
 
-    muxed.on('stream', this.handler)
-    this._readDiscovery()
+    if (!this.client.discovery._isStarted) {
+      log('reading discovery, but tossing data since it\'s not enabled')
+      return
+    }
+
+    log('reading discovery')
+    message.ids
+      .map(id => {
+        const pi = new PeerInfo(new PeerId(id))
+        if (pi.id.toB58String() === this.client.id.toB58String()) return
+        pi.multiaddrs.add(baseAddr.encapsulate('/p2p-stardust/p2p/' + pi.id.toB58String()))
+
+        return pi
+      })
+      .filter(Boolean)
+      .forEach(pi => this.client.discovery.emit('peer', pi))
+
+    setTimeout(() => this._onDiscovery(), 100) // cooldown
   }
 
   async _dial (addr) {
-    if (!this.connected) { throw new Error('Server not online!') }
+    if (!this.isConnected) { throw new Error('Server not online!') }
 
     const id = addr.getPeerId()
-    const _id = ID.createFromB58String(id)._id
+    const _id = PeerId.createFromB58String(id)._id
```

Comment:
We can use the getter to avoid using the private property `_id`
```suggestion
    const _id = PeerId.createFromB58String(id).id
```